File: Cloud Mail Samples/Gmail/java/gmail.java
/*
 * Cloud Mail 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Mail in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudmail
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import cloudmail.*;

import java.nio.charset.StandardCharsets;

public class gmail extends ConsoleDemo {
  static Gmail gmail1 = new Gmail();
  static String command; // user's command
  static String[] argument; // arguments following command
  static boolean isListingEmails = false;


  public static void main(String[] args) {
    try{
      gmail1.getOAuth().setClientId("922417066392-f9fomlr35hffi9gkq5te61nr47rtk00q.apps.googleusercontent.com");
      gmail1.getOAuth().setClientSecret("o8tCrbErRHjAh42whuvXGW0R");
      gmail1.getOAuth().setServerAuthURL("https://accounts.google.com/o/oauth2/auth");
      gmail1.getOAuth().setServerTokenURL("https://accounts.google.com/o/oauth2/token");
      gmail1.getOAuth().setAuthorizationScope("https://mail.google.com/");
      gmail1.authorize();

      gmail1.addGmailEventListener(new DefaultGmailEventListener() {
        public void messageInfo(GmailMessageInfoEvent e) {
          if(isListingEmails) {
            System.out.println(e.id);
          }
        }
      });
      DisplayMenu();
      while(true) {
        command = prompt("gmail", ">");
        argument = null;
        if(command.length() == 0)
          command = "?";
        
        argument = command.split("\\s");
        if (argument.length == 0)
          continue;

        switch (argument[0].charAt(0)) {
          case 'l':
            isListingEmails = true;
            gmail1.config("ResponseType=4");
            gmail1.listMessages("", "");
            break;
          case 'v':
            isListingEmails = false;
            gmail1.config("ResponseType=0");
            gmail1.retrieveMessageRaw(argument[1].trim());
            String s = new String(gmail1.getMessage(), StandardCharsets.UTF_8);
            System.out.println(s);
            break;
          case 'c':
            messageCompose();
            break;
          case '?':
            DisplayMenu();
            break;
          case 'q':
            System.exit(0);
          default: // allow user to enter only the number of the message they
            // want to view
            try {
              isListingEmails = false;
              gmail1.config("ResponseType=0");
              gmail1.retrieveMessageRaw(command);
            } catch (NumberFormatException e) {
              System.out.println("Bad command / Not implemented in demo.");
            }
        }
      }
    } catch (Exception e) {
      System.out.println(e.getMessage());
    }
  }

  public static void DisplayMenu() {
    System.out.println("Gmail commands");
    System.out.println("l                   list next 100 messages");
    System.out.println("v <message id>      view the content of selected message");
    System.out.println("c                   compose a message");
    System.out.println("?                   display options");
    System.out.println("q                   quit");
  }

  public static void messageCompose() {
    try{
      String sendTo = prompt("Send To", ":");
      if (!sendTo.isEmpty())
        gmail1.setMessageTo(sendTo);

      String cc = prompt("CC", ":");
      if (!sendTo.isEmpty())
        gmail1.setMessageCc(cc);


      String bcc = prompt("BCC", ":");
      if (!sendTo.isEmpty())
        gmail1.setMessageBcc(bcc);

      String subject = prompt("Subject", ":");
      if (!sendTo.isEmpty())
        gmail1.setMessageSubject(subject);

      String messageBodyType = prompt("Message Body Type (text/html)", ":");
      while(!messageBodyType.equalsIgnoreCase("text") && !messageBodyType.equalsIgnoreCase("html")) {
        System.out.println("Please enter either text or html as a value!");
        messageBodyType = input();
      }
      gmail1.setMessageBodyContentType(messageBodyType);

      String content = prompt("Message Content", ":");
      if (!sendTo.isEmpty())
        gmail1.setMessageBodyContent(content);

      String attachment = prompt("Attachments(semicolon separated)", ":");
      if (!sendTo.isEmpty() && attachment.length() > 0) {
        String[] array = attachment.split(";");
        boolean flag = true;
        for(int i = 0; i < array.length; i++) {
          File file = new File(array[i].trim());
          if(!file.exists()) {
            flag = false;
            break;
          }
        }
        if(flag)
          gmail1.setMessageAttachments(attachment);
        else
          System.out.println("Attachments invalid. Sending no attachments!");
        
      }
      gmail1.sendMail();
      System.out.println("Message sent!");

    } catch(Exception e) {
      displayError(e);
    }

  }
}

class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "]" + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof CloudMailException) {
      System.out.print(" (" + ((CloudMailException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }

  /**
   * Takes a list of switch arguments or name-value arguments and turns it into a map.
   */
  static java.util.Map<String, String> parseArgs(String[] args) {
    java.util.Map<String, String> map = new java.util.HashMap<String, String>();
    
    for (int i = 0; i < args.length; i++) {
      // Add a key to the map for each argument.
      if (args[i].startsWith("-")) {
        // If the next argument does NOT start with a "-" then it is a value.
        if (i + 1 < args.length && !args[i + 1].startsWith("-")) {
          // Save the value and skip the next entry in the list of arguments.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), args[i + 1]);
          i++;
        } else {
          // If the next argument starts with a "-", then we assume the current one is a switch.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), "");
        }
      } else {
        // If the argument does not start with a "-", store the argument based on the index.
        map.put(Integer.toString(i), args[i].toLowerCase());
      }
    }
    return map;
  }
}

File: Cloud Mail Samples/Gmail/python/gmail.py
# 
# Cloud Mail 2024 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of Cloud Mail in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/cloudmail
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from cloudmail import *

input = sys.hexversion < 0x03000000 and raw_input or input


def ensureArg(args, prompt, index):
    if len(args) <= index:
        while len(args) <= index:
            args.append(None)
        args[index] = input(prompt)
    elif args[index] is None:
        args[index] = input(prompt)


import traceback

gmail1 = GMail()
command = ""
arguments = []
isListingEmails = False

def fire_message_info(e):
  if isListingEmails:
    print(e.id)

def display_error(e):
  print("Error", end = '')
  if isinstance(e, CloudMailError):
    print(" (" + str(e.code) + ")", end = '')
  print(": " + str(e.message))
  traceback.print_exc()

def display_menu():
  print("Gmail Commands:")
  print("l                   list next 100 messages")
  print("v <message id>      view the content of selected message")
  print("c                   compose a message")
  print("?                   display options")
  print("q                   quit")

def message_compose():
  try:
    sendTo = input("Send To: ")
    if len(sendTo) != 0:
      gmail1.set_message_to(sendTo)

    cc = input("CC: ")
    if len(cc) != 0:
      gmail1.set_message_cc(cc)

    bcc = input("BCC: ")
    if len(bcc) != 0:
      gmail1.set_message_bcc(bcc)

    subject = input("Subject: ")
    if len(subject) != 0:
      gmail1.set_message_subject(subject)

    messageBodyType = input("Message Body Type (text/html): ")
    while messageBodyType.casefold() != "text" and messageBodyType.casefold() != "html":
      print("Please enter either text or html as a value!")
      messageBodyType = input("Message Body Type (text/html): ")
    gmail1.set_message_body_content_type(messageBodyType)

    content = input("Message Content: ")
    if len(sendTo) != 0:
      gmail1.set_message_body_content(content)

    attachment = input("Attachments (semicolon separated): ")
    if len(sendTo) != 0 and len(attachment) > 0:
      attachmentList = attachment.split(";")
      for attachmentItem in attachmentList:
        if os.path.isfile(attachmentItem.strip()) is False:
          print("Attachments invalid.  Sending no attachments!")
          break # skips subsequent else clause
      else:
        gmail1.set_message_attachments(attachment)

    gmail1.send_mail()
    print("Message sent!")
  except Exception as e:
    display_error(e)

try:
  gmail1.set_o_auth_client_id("922417066392-f9fomlr35hffi9gkq5te61nr47rtk00q.apps.googleusercontent.com")
  gmail1.set_o_auth_client_secret("o8tCrbErRHjAh42whuvXGW0R")
  gmail1.set_o_auth_server_auth_url("https://accounts.google.com/o/oauth2/auth")
  gmail1.set_o_auth_server_token_url("https://accounts.google.com/o/oauth2/token")
  gmail1.set_o_auth_authorization_scope("https://mail.google.com/")
  gmail1.authorize()

  gmail1.on_message_info = fire_message_info

  display_menu()

  while (True):
    command = input("gmail> ")
    argument = None
    if len(command) == 0:
      command = "?"

    argument = command.split()
    if len(argument) == 0:
      continue

    if argument[0][0] == 'l':
      isListingEmails = True
      gmail1.config("ResponseType = 4")
      gmail1.list_messages("", "")
    elif argument[0][0] == 'v':
      isListingEmails = False
      gmail1.config("ResponseType = 0")
      gmail1.fetch_message_raw(argument[1].strip())
      print(gmail1.get_message())
    elif argument[0][0] == 'c':
      message_compose()
    elif argument[0][0] == '?':
      display_menu()
    elif argument[0][0] == 'q':
      sys.exit(0)

except Exception as e:
  display_error(e)



